### ISSUE TITLE
Update Python version from 3.6.7 to 3.9.10

### ISSUE THREAD

**Issue by jcfr** [Slicer/Slicer#5014]
This was initially discussed in https://github.com/Slicer/Slicer/issues/4859

The following PR would need to be finalized:

- <s>[ ] python-cmake-buildsystem/python-cmake-buildsystem#264</s>
- [x] python-cmake-buildsystem/python-cmake-buildsystem#267
- [x] When Slicer's Python is upgraded then post a comment to MONAILabel issue https://github.com/Project-MONAI/MONAILabel/pull/378 (so that they can remove Python-3.6 support from MONAILabel)

**Comment by jamesobutler** [comment]
Python packages that have dropped Python 3.6 support:
- Numpy >=1.20.0
- Scipy >=1.6.0

We won't be able to upgrade to these new versions until the Python version included in Slicer is upgraded.

**Comment by jamesobutler** [comment]
Additionally the following python packages not included in Slicer by default have dropped Python 3.6 support (mostly per [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html)):
- scikit-image >=0.18.0
- Matplotlib >= 3.4.0
- IPython >= 7.17.0
- pandas >= 1.2.0

**Comment by jcfr** [comment]
> The following PR would need to be finalized

Will have a look in early January

**Comment by lassoan** [comment]
Thank you @jamesobutler for helping with this. I've already started to encounter issues with lack of packages for Python-3.6, so it is time to move on.

**Comment by lassoan** [comment]
Python-3.6 is starting to become a problem.

For example, we had to stop updating SlicerJupyter, because recent xeus-python versions require Python-3.7. (see https://github.com/Slicer/SlicerJupyter/issues/62)

**Comment by lassoan** [comment]
Yet another important reason to update Python in Slicer ASAP: https://github.com/Project-MONAI/MONAILabel/pull/378

**Comment by jamesobutler** [comment]
GitPython >= 3.1.22 has dropped support for Python 3.6 as well

**Comment by jamesobutler** [comment]
Note that per [NEP29](https://numpy.org/neps/nep-0029-deprecation_policy.html#) projects have begun to drop Python 3.7 support. 

For example the following projects have dropped Python 3.7 support already for their upcoming releases.
- Numpy >= 1.22.0
- Scipy >= 1.8.0

Also to note that Python 3.10 has been released as of Oct. 4th 2021

**Comment by fbordignon** [comment]
Hi all, I am trying to make progress on this issue with a PR to python-cmake-buildsystem https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/281
I am available to discuss the issues and test a Slicer build with python 3.10

**Comment by jamesobutler** [comment]
@jcfr do you have to plans to finalize those PRs to add more python version support? You are the maintainer of that repo so will need your help there.

**Comment by jamesobutler** [comment]
Python 3.6 goes End-Of-Life in 42 days (Dec 23rd 2021). https://endoflife.date/python

**Comment by fbordignon** [comment]
@jamesobutler I am trying to do some work on this but my skills at cmake are not sufficient. I've commented on the [PR](https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/281) I've opened. I think we are close, but it is taking a lot of time for me to figure out stuff.

**Comment by fbordignon** [comment]
![image](https://user-images.githubusercontent.com/4790289/141456908-77fa8a58-019a-43a1-94fa-2bf704bf52ce.png)

Still have a problem with SimpleITK but it went all the way to the CPack and generated a package. I had to do some dirty manual workarounds that I plan to work on and make it less dirty. At least now I know that it is possible 👍

**Comment by fbordignon** [comment]
I've managed to solve the issue with SimpleITK and it is working now.

However, I am concerned about the fact that many popular libs don't have a pypi prebuilt wheel for python 3.10
Example: scikit-image, matplotlib, pyside2

Hence, I believe the users would benefit the most from an update to python 3.9, so they would have more packages available to install directly via pip without the need to build the package, which is problematic in some cases.

I am ready to abandon the idea of an update to 3.10 for now, in favor of updating to 3.9 at least until more support is added to python 3.10.

**Comment by jamesobutler** [comment]
Are there any python packages in Slicer core that don’t have Python 3.10 wheels available on PyPI?

It looks like in some cases that packages are about to release updates where Python 3.10 wheels will be available. For instance matplotlib’s 3.5.0 milestone is almost complete and they have Python 3.10 wheels available for their release candidate as seen https://pypi.org/project/matplotlib/3.5.0rc1/#files. If wheels are available here soon, wouldn’t want to waste effort to only do Python 3.9 support that would then become Python 3.10 in Slicer in just a few weeks.

**Comment by lassoan** [comment]
Indeed, wheels are only available for cp37, cp38, cp39 for both https://pypi.org/project/matplotlib/#files and https://pypi.org/project/scikit-image/#files. Not that we care about any particular package, but such basic packages missing indicate that Python-3.10 is just not ready for users yet. This article lists many other issues with Python-3.10: https://pythonspeed.com/articles/switch-python-3.10/

Since Slicer is a consumer of Python packages, it should switch to a Python version when that Python is ready for Python _users_. Python-3.10 looks more like it is ready for Python package _developers_ to start testing their packages. It is unfortunate that [Python website](https://www.python.org/downloads/) already advertises Python-3.10 as if it was ready for end users, while it is clearly not.

Probably it'll take just a few months for most packages to get updated for Python-3.10, but many will take more time, so it could make sense for Slicer to update to Python-3.9 now and move on to Python-3.10 in about 2 years (unless there are specific reasons to move earlier).

**Comment by fbordignon** [comment]
> Are there any python packages in Slicer core that don’t have Python 3.10 wheels available on PyPI?

Not that I've noticed.

> If wheels are available here soon, wouldn’t want to waste effort to only do Python 3.9 support that would then become Python 3.10 in Slicer in just a few weeks.

Good point, I agree with it but I am still reluctant. I would bet it is a safer path to go with 3.9 since we have almost 4 years until the end of life of 3.9. I can help with the upgrade to 3.10 if more wheels are made available.

**Comment by jamesobutler** [comment]
> Python-3.10 looks more like it is ready for Python package developers to start testing their packages. It is unfortunate that Python website already advertises Python-3.10 as if it was ready for end users, while it is clearly not.

It appears that it is just always the situation of people wait until something is publicly released to update to use it. So if Python3 3.10 was released later, it would just result in issues later on as people would still not take advantage of it earlier.

Considering that article was written a couple days after the release of Python 3.10 I would guess that many of its points are already outdated since it is almost a month and half after the release.

**Comment by jamesobutler** [comment]
> > Are there any python packages in Slicer core that don’t have Python 3.10 wheels available on PyPI?
> 
> Not that I've noticed.

That's great. Note that you can use https://github.com/Slicer/Slicer/blob/master/Utilities/Scripts/python_package_updater.py to help updating all the version hashes that correspond to the new python version.

**Comment by lassoan** [comment]
The next 1-2 years (age 1-3) will be the best years of Python-3.9: already stable but still widely used and maintained. If we developed a Python library then of course we would need to start working on Python-3.10 compatibility ASAP. But as application developers, it would not make sense for us to move to Python-3.10 now, unless there is some extremely important feature for us in 3.10 that would make it worth to take all the risk and inconvenience of being early adopters. We need to give at least a half year to a year for library developers to upgrade and stabilize their stuff.

Unless anybody knows about some game changer feature in 3.10 or any specific insight into Python politics or strategies, we should be still on 3.9 for at least a year.

**Comment by fbordignon** [comment]
I agree with Andras. Let's wait for more developers to jump on board of the 3.10 train and we can update to 3.10 in a year or so.
The upgrade should be smoother if we keep upgrading to each new version. I will open a PR on python-cmake-buildsystem for python 3.9.8.

**Comment by jamesobutler** [comment]
Yeah you can issue a PR to pythonCmakeBuildSystem for both 3.9 and 3.10 and then later we can just update Slicer to configure it for the newer version and it’ll all be ready to go without additional updates.

**Comment by fbordignon** [comment]
Guys, I am almost done with the commits to complete the update to python 3.9.8 but I need help with one issue.

Python 3.8+ changed a little bit the behavior of dll loading on windows see: https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
TL;DR; dll search path does not include the PATH env variable anymore, we need to explicitly call os.add_dll_directory passing relevant dirs

Two places that are suffering with this:
1 - python needs the openssl dlls for the ssl package which are loaded using the path during the install step of the python-* targets (example: python-numpy). There is no easy way to set the environment without calling os.add_dll_directory on the running python session. The only way I can think to solve this is to copy the openssl dlls to the python-install/bin for it to see the dlls and use them at the python-* install steps. At the final slicer package, the openssl dlls are already going to be on the same folder of python (slicer/bin) and there will be no problem.

Another option I am investigating is to add the following script to PythonSlicerLauncherSettings.ini using PYTHONSTARTUP=./add_dll_directories_windows.py
```python
import sys
if sys.version_info[1] >= 8 and sys.platform == 'win32':
    import os
    libpath = os.getenv("LIBRARYPATH")
    for path in libpath.split(";"):
        os.add_dll_directory(path)
```
add 



2- SimpleITK install. There is a python command to install it on python where SimpleITK depends on ITK, DCM, etc... So the simplest solution would be to create an environment variable like SIMPLE_ITK_PYTHON_LIBS and add calls to os.add_dll_directory to all the items on the list at SimpleITK/Wrapping/Python/SimpleITK/__init__.py. Similar to what I proposed above.

On the final package, there is no need to any of this, because everything is on the same folder and it works.

Please advise if you think there are problems with these approaches.

I've manually overcome this issues and did a few tests. Everything seems to be working
![image](https://user-images.githubusercontent.com/4790289/141767705-5ff253c1-0886-4603-8509-49caf8b9487b.png)

**Comment by jamesobutler** [comment]
Because previously discussed in https://github.com/Slicer/Slicer/issues/5014#issuecomment-967454827 about wheels for Python 3.10, I’ll provide an update here that matplotlib 3.5.0 is out now with Python 3.10 wheels https://pypi.org/project/matplotlib/3.5.0/#files. 

Pyside2 for Qt5 doesn’t have a Python 3.10 wheel, but Pyside6 for Qt6 does have Python 3.10 wheel https://pypi.org/project/PySide6/6.2.1/#files. Could potentially be used when Slicer goes to Qt6

**Comment by jamesobutler** [comment]
Not that it changes anything immediately, but

Python 3.6 goes End-Of-Life tomorrow (Dec 23rd 2021). https://endoflife.date/python

**Comment by fbordignon** [comment]
Thanks for the reminder.
We (LTrace) are putting python 3.9.9 in production on GeoSlicer for the next version due in 3 weeks as we found no issues with the current PR and our users need it desperately. Some libs we are working with does not support 3.6 anymore (pyqtgraph) and others never supported it (porespy despite having a wheel for any version in pypi).

**Comment by lassoan** [comment]
@jcfr is there anything that you would like to implement or test before integrating this?

**Comment by jcfr** [comment]
* Packaging of Slicer on our three platforms 
* Updated CI (GitHub action) on python-cmake-buildsystem

**Comment by Bugsbunny2917** [comment]
> Guys, I am almost done with the commits to complete the update to python 3.9.8 but I need help with one issue.
> 
> Python 3.8+ changed a little bit the behavior of dll loading on windows see: https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew TL;DR; dll search path does not include the PATH env variable anymore, we need to explicitly call os.add_dll_directory passing relevant dirs
> 
> Two places that are suffering with this: 1 - python needs the openssl dlls for the ssl package which are loaded using the path during the install step of the python-* targets (example: python-numpy). There is no easy way to set the environment without calling os.add_dll_directory on the running python session. The only way I can think to solve this is to copy the openssl dlls to the python-install/bin for it to see the dlls and use them at the python-* install steps. At the final slicer package, the openssl dlls are already going to be on the same folder of python (slicer/bin) and there will be no problem.
> 
> Another option I am investigating is to add the following script to PythonSlicerLauncherSettings.ini using PYTHONSTARTUP=./add_dll_directories_windows.py
> 
> ```python
> import sys
> 
> if sys.version_info[1] >= 8 and sys.platform == 'win32':
>     import os
>     libpath = os.getenv("LibraryPaths")
>     
>     if libpath is not None:
>         for path in libpath.split(";"):
>             p = os.path.abspath(path)
>             if os.path.exists(p):
>                 os.add_dll_directory(p)
> ```
> 
> modify the .ini script to add LibraryPaths to path variables
> 
> ```
> [Environment]
> additionalPathVariables=PYTHONPATH,LibraryPaths
> 
> ...
> 
> [EnvironmentVariables]
> ...
> PYTHONSTARTUP=./add_dll_directories_windows.py
> ```
> 
> 2- SimpleITK install. There is a python command to install it on python where SimpleITK depends on ITK, DCM, etc... So the simplest solution would be to create an environment variable like SIMPLE_ITK_PYTHON_LIBS and add calls to os.add_dll_directory to all the items on the list at SimpleITK/Wrapping/Python/SimpleITK/**init**.py. Similar to what I proposed above. EDIT: Actually, I believe that the script in 1 can fix number 2 also. I am testing it now.
> 
> Please advise if you think there are problems with these approaches.
> 
> I've manually overcome this issues and did a few tests. Everything seems to be working ![image](https://user-images.githubusercontent.com/4790289/141767705-5ff253c1-0886-4603-8509-49caf8b9487b.png)

Hello sir, i'm working in firm where we were using 3D Slicer, some libraries requires Python 3.9, if you would send the updated slicer, we can use that also we can tested it. It would be our great pleasure to contribute to it.

**Comment by fbordignon** [comment]
@Bugsbunny2917 you can build 3D Slicer from the branch of https://github.com/Slicer/Slicer/pull/6026 
It is working with python 3.9.9 and should be integrated soon.

**Comment by ningw-git** [comment]
Hello, does slicer have a version that uses python3.8? I'm using a ROKAE robot recently, and they provided a program and library file based on python3.8 (similar to: robot.cp38-win_amd64.pyd). I want to use this robot in slicer, so please help.

**Comment by lassoan** [comment]
Please create a new topic at https://discourse.slicer.org to discuss this. If we tell how to connect to a ROKAE robot here then others who want to do something similar will not find our answer.

**Comment by ningw-git** [comment]
thanks，I have post a new topic:
&nbsp;does slicer have a version that uses python3.8?
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Slicer/Slicer"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月9日(星期三) 晚上7:24
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Slicer/Slicer] Update Python version from 3.6.7 to 3.9.10 (#5014)





 
Please create a new topic at https://discourse.slicer.org to discuss this. If we tell how to connect to a ROKAE robot here then others who want to do something similar will not find our answer.
